### PR TITLE
Separate backend

### PR DESCRIPTION
**Pull request by robdockins**

### CODE REVIEW

File: semmc-aarch32/test/TestMain.hs [hunk 1]
```diff
@@ -55,8 +55,8 @@ data Flag = N | Z | C | V | Q
   deriving (Show, Eq)
 
 -- | Given a list of flags initializes a CPSR set to user mode
-mkCPSR :: [Flag] -> Word32
-mkCPSR flags = foldr (.|.) 16 [n,z,c,v,q]
+_mkCPSR :: [Flag] -> Word32
```
Comment (RyanGlScott) [inline semmc-aarch32/test/TestMain.hs:58]: This is unrelated to the main thrust of the PR, but is there any reason to have this `mkCPSR` function (and associated `Flag` data type) here at all? It's certainly not used in the module, and I don't see why it should belong in the test suite driver. I'd favor just removing it.

File: semmc-arm/tools/Main.hs [hunk 2]
```diff
@@ -55,8 +55,8 @@ data Flag = N | Z | C | V | Q
   deriving (Show, Eq)
 
 -- | Given a list of flags initializes a CPSR set to user mode
-mkCPSR :: [Flag] -> Word32
-mkCPSR flags = foldr (.|.) 16 [n,z,c,v,q]
+_mkCPSR :: [Flag] -> Word32
```
Comment (RyanGlScott) [inline semmc-arm/tools/Main.hs:58]: Similarly, I'd favor removing `mkCPSR` and `Flag` from here.

File: semmc-learning/src/SemMC/Stochastic/Initialize.hs [hunk 3]
```diff
@@ -82,10 +85,10 @@ data Config arch =
          }
 
 -- | Synthesis environment.
-data SynEnv t solver fs arch =
-  SynEnv { seFormulas :: STM.TVar (MapF.MapF ((A.Opcode arch) (A.Operand arch)) (F.ParameterizedFormula (Sym t solver fs) arch))
+data SynEnv solver t fs arch =
```
Comment (RyanGlScott) [inline semmc-learning/src/SemMC/Stochastic/Initialize.hs:88]: What is the motivation for swapping the order of the `solver` and `t` type parameters here? It's the cause of a fair bit of downstream code churn, but I'm not sure why it's necessary.

File: semmc-synthesis/src/SemMC/Synthesis/Cegis.hs [hunk 4]
```diff
@@ -55,21 +55,22 @@ data CegisResult sym arch = CegisUnmatchable [ConcreteTest sym arch]
...
+       -> [T.TemplatedInstructionFormula (WE.ExprBuilder t st fs) arch]
        -- ^ The trial instructions.
-       -> Formula (CBO.OnlineBackend t solver fs) arch
+       -> Formula (WE.ExprBuilder t st fs) arch
        -- ^ A formula representing the sequence of trial instructions.
-       -> Cegis (CBO.OnlineBackend t solver fs) arch (CegisResult (CBO.OnlineBackend t solver fs) arch)
-cegis' trial trialFormula = do
+       -> Cegis (WE.ExprBuilder t st fs) arch (CegisResult (WE.ExprBuilder t st fs) arch)
+cegis' bak trial trialFormula = do
   sym <- askSym
```
Comment (RyanGlScott) [inline semmc-synthesis/src/SemMC/Synthesis/Cegis.hs:71]: Could this use `askBackend` instead? That way, you wouldn't need to change the type of `cegis'` to add an explicit `bak` argument. (There are various other functions that now take an additional `bak` argument for which the same question applies.)

File: semmc-synthesis/src/SemMC/Synthesis/Cegis/LLVMMem.hs [hunk 5]
```diff
@@ -59,7 +60,12 @@ instance T.HasMemExpr MemM
 
 instance T.HasSym MemM
   where
-    askSym = MemM $ memSym <$> get
+    askSym = MemM $
+      do SomeBackend bak <- memBackend <$> get
+         return (B.backendGetSym bak)
+      
+askBackend :: MemM sym w (SomeBackend sym)
+askBackend = MemM$ memBackend <$> get
```
Comment (RyanGlScott) [inline semmc-synthesis/src/SemMC/Synthesis/Cegis/LLVMMem.hs:68]: Style nitpick:

```suggestion
askBackend = MemM $ memBackend <$> get
```

File: semmc-synthesis/src/SemMC/Synthesis/Cegis/Types.hs [hunk 6]
```diff
@@ -63,7 +66,7 @@ import qualified SemMC.Synthesis.Template as T
 
 -- | Parameters given to call cegis.
 data CegisParams sym arch =
-  CegisParams { cpSym :: sym
+  CegisParams { cpBackend :: SomeBackend sym
               -- ^ The symbolic expression builder.
```
Comment (RyanGlScott) [inline semmc-synthesis/src/SemMC/Synthesis/Cegis/Types.hs:70]: Now that this is a backend, this comment is out of date, I believe.

File: semmc/src/SemMC/Architecture.hs [hunk 7]
```diff
@@ -290,8 +291,9 @@ taggedExprImmediate _ _ = error "Other tagged operand"
 -- We need to pass the return type 'BaseTypeRepr' in so that we can know at the
 -- call site that the expression produced by the evaluator is correctly-typed.
 data Evaluator arch t st fs =
```
Comment (RyanGlScott) [inline semmc/src/SemMC/Architecture.hs:293]: Unrelated to the PR at hand, but this probably should be a newtype.

File: semmc-ppc/tests/Main.hs [hunk 8]
```diff
@@ -156,36 +160,38 @@ mkMemRI n i = D.Memri (D.MemRI (Just (D.GPR n)) i)
...
+               (WPO.OnlineSolver solver, sym ~ WEB.ExprBuilder t st fs, CB.IsSymInterface sym)
             => [(String, BS8.ByteString)]
             -> [(Some (D.Opcode D.Operand), BS8.ByteString)]
-            -> sym
+            -> CBO.OnlineBackend solver t st fs
             -> IO (MapF.MapF (D.Opcode D.Operand) (SF.ParameterizedFormula sym PPC64.PPC),
                    SS.SynthesisEnvironment sym PPC64.PPC)
-loadBaseSet funcs ops sym = do
+loadBaseSet funcs ops bak = do
+  let sym = (CB.backendGetSym bak)
```
Comment (RyanGlScott) [inline semmc-ppc/tests/Main.hs:188]: Style nitpick:

```suggestion
  let sym = CB.backendGetSym bak
```

File: semmc-aarch32/test/TestMain.hs (hunk 1 above)
Comment (robdockins) [reply semmc-aarch32/test/TestMain.hs:58]: I'm OK with that, I don't really know what these functions are/were for.

File: semmc-learning/src/SemMC/Stochastic/Initialize.hs (hunk 3 above)
Comment (robdockins) [reply semmc-learning/src/SemMC/Stochastic/Initialize.hs:88]: I'm trying to keep things in sync with the parameter order of the `OnlineBackend`.  I think it's really confusing when things like this don't match.

File: semmc-synthesis/src/SemMC/Synthesis/Cegis.hs (hunk 4 above)
Comment (robdockins) [reply semmc-synthesis/src/SemMC/Synthesis/Cegis.hs:71]: I don't think so: the `askBackend` gives a `SomeBackend`, which existentially closes over the type. In some places (including here, I'm pretty sure) we need to actually know that we have an `OnlineBackend`. There are other ways, maybe, to achieve this, but this seemed like the easiest to me.

File: semmc-learning/src/SemMC/Stochastic/Initialize.hs (hunk 3 above)
Comment (travitch) [reply semmc-learning/src/SemMC/Stochastic/Initialize.hs:88]: I'm aligned. I've been trying to fix that kind of thing myself

**Review by travitch — APPROVED** [review]

**Review by RyanGlScott — APPROVED** [review]

**Pull request merged**